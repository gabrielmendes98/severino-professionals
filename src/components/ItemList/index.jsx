import PropTypes from 'prop-types';
import ListItem from '@material-ui/core/ListItem';
import Paper from '@material-ui/core/Paper';
import List from '@material-ui/core/List';
import EditIcon from '@material-ui/icons/Edit';
import DeleteIcon from '@material-ui/icons/Delete';
import showDeleteModal from 'components/Modal/showModalCommons/showDeleteModal';
import IconButton from 'components/IconButton';
import {
  Wrapper,
  Container,
  ItemTemplateContainer,
  ItemTemplateActions,
} from './style';

const ItemList = ({ items, deleteItem, editItem, id, ItemTemplate }) => (
  <Paper component={Container}>
    <List component={Wrapper} role="list">
      {items?.map((item, index) => (
        <ListItem
          key={item.id}
          divider={index !== items.length - 1}
          role="listitem"
        >
          <ItemTemplateContainer>
            <ItemTemplate item={item} />
          </ItemTemplateContainer>

          <ItemTemplateActions>
            {editItem && (
              <IconButton
                color="primary"
                tooltip="Editar"
                onClick={() => editItem(item)}
                // eslint-disable-next-line prefer-template
                id={'edit-' + id + '-' + index}
              >
                <EditIcon />
              </IconButton>
            )}
            <IconButton
              tooltip="Deletar"
              onClick={() =>
                showDeleteModal({ onConfirm: () => deleteItem(item.id) })
              }
              // eslint-disable-next-line prefer-template
              id={'delete-' + id + '-' + index}
            >
              <DeleteIcon />
            </IconButton>
          </ItemTemplateActions>
        </ListItem>
      ))}
    </List>
  </Paper>
);

ItemList.propTypes = {
  items: PropTypes.array.isRequired,
  deleteItem: PropTypes.func.isRequired,
  editItem: PropTypes.func,
  id: PropTypes.string.isRequired,
  ItemTemplate: PropTypes.any.isRequired,
};

export default ItemList;
