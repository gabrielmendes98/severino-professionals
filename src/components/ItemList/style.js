import styled from 'styled-components';

export const Container = styled.div`
  width: 100%;
`;

export const Wrapper = styled.ul`
  width: 100%;
`;

export const ItemTemplateContainer = styled.div`
  flex: 1;
`;

export const ItemTemplateActions = styled.div``;
