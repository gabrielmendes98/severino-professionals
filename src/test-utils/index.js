/* eslint-disable react/prop-types */
import { render, configure, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Form, Formik } from 'formik';
import MainProvider from 'commons/providers/MainProvider';

configure({ testIdAttribute: 'id' });

const Wrapper = ({ children }) => (
  <>
    <MainProvider>{children}</MainProvider>
    <div id="modal"></div>
  </>
);
const RouteWrapper = ({ children }) => (
  <BrowserRouter>
    <Wrapper>{children}</Wrapper>
  </BrowserRouter>
);
const FormikWrapper = ({ children, initialValues, ...config }) => (
  <Formik initialValues={initialValues} {...config}>
    <Form>{children}</Form>
  </Formik>
);

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options });

const renderWithRouter = (ui, { route = '/', ...options } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: RouteWrapper, ...options });
};

const asFragment = ui => {
  const { asFragment } = customRender(ui);
  return asFragment(ui);
};

const toMatchSnapshot = (Component, { useRouter = false } = {}) => {
  const wrapper = asFragment(
    useRouter ? <BrowserRouter>{Component}</BrowserRouter> : Component,
  );
  expect(wrapper).toMatchSnapshot();
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export {
  customRender as render,
  screen,
  toMatchSnapshot,
  asFragment,
  renderWithRouter,
  FormikWrapper,
};
