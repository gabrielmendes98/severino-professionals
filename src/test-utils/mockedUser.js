const mockedUser = {
  id: '9c6cb22e-f284-4ca4-9a88-d64ece7b896c',
  createdAt: '2021-09-02T03:55:52.338Z',
  updatedAt: '2021-09-02T03:55:52.338Z',
  email: 'test@test.com',
  name: 'Testerson da Silva',
  description: null,
  avatarUrl: 'www.avatarurlfake.com',
  phone: '(11) 11111-1112',
  state: 'Minas Gerais',
  city: '√Åguas Formosas',
  hasWhatsapp: false,
};
export default mockedUser;
