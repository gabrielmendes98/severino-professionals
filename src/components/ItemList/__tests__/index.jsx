import { toMatchSnapshot, render, screen, userEvent } from 'test-utils';
import ItemList from '..';

const data = [
  {
    id: 'b2cbb858-7a56-4ae1-bdf9-eaa62a417257',
    createdAt: '2021-10-11T12:45:05.685Z',
    updateAt: '2021-10-11T12:45:05.685Z',
    jobId: '4165edeb-19ab-4668-8ddd-009b38205e0a',
    profileId: 'f005dc18-698c-45eb-9bc2-e470a94abc94',
    role: 'Engenheiro de Software',
    company: 'TQI',
    city: 'UberlÃ¢ndia',
    state: 'Minas Gerais',
    startDate: '2020-10-11T12:44:59.000Z',
    endDate: '2021-10-11T12:44:59.000Z',
    job: {
      id: '4165edeb-19ab-4668-8ddd-009b38205e0a',
      type: 'FULL_TIME',
      description: 'Tempo integral',
    },
  },
  {
    id: 'd5b37c62-ee40-4627-aa9e-7daa4a09e46a',
    createdAt: '2021-10-11T13:00:19.418Z',
    updateAt: '2021-10-11T13:00:19.418Z',
    jobId: '4165edeb-19ab-4668-8ddd-009b38205e0a',
    profileId: 'f005dc18-698c-45eb-9bc2-e470a94abc94',
    role: 'Pedreiro',
    company: '',
    city: 'Jeceaba',
    state: 'Minas Gerais',
    startDate: '2020-10-11T12:59:37.000Z',
    endDate: '2022-10-11T12:59:37.000Z',
    job: {
      id: '4165edeb-19ab-4668-8ddd-009b38205e0a',
      type: 'FULL_TIME',
      description: 'Tempo integral',
    },
  },
];

// eslint-disable-next-line react/prop-types
const ItemTemplate = ({ item }) => <div>{item.role}</div>;

it('should match snapshot', () => {
  toMatchSnapshot(
    <ItemList
      items={data}
      deleteItem={() => {}}
      editItem={() => {}}
      id="test"
      ItemTemplate={ItemTemplate}
    />,
  );
});

it('should call functions on button click', async () => {
  const deleteItem = jest.fn();
  const editItem = jest.fn();
  render(
    <ItemList
      items={data}
      deleteItem={deleteItem}
      editItem={editItem}
      id="test"
      ItemTemplate={ItemTemplate}
    />,
  );

  userEvent.click(screen.getByTestId('edit-test-0'));
  expect(editItem).toHaveBeenCalledTimes(1);

  userEvent.click(screen.getByTestId('delete-test-0'));
  userEvent.click(await screen.findByTestId('modal-confirm'));
  expect(deleteItem).toHaveBeenCalledTimes(1);
});
