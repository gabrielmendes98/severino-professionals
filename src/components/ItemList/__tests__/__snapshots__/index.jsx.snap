// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  .c3 {
  margin-top: 0px;
  margin-left: 0px;
  margin-right: 0px;
  margin-bottom: 0px;
}

.c0 {
  width: 100%;
}

.c1 {
  width: 100%;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div
    class="c0 MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <ul
      class="c1 MuiList-root MuiList-padding"
      role="list"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-divider"
        role="listitem"
      >
        <div
          class="c2"
        >
          <div>
            Engenheiro de Software
          </div>
        </div>
        <div
          class=""
        >
          <button
            aria-label="Editar"
            class="MuiButtonBase-root MuiIconButton-root c3 MuiIconButton-colorPrimary"
            id="edit-test-0"
            tabindex="0"
            title="Editar"
            type="button"
            variant="contained"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Deletar"
            class="MuiButtonBase-root MuiIconButton-root c3 MuiIconButton-colorPrimary"
            id="delete-test-0"
            tabindex="0"
            title="Deletar"
            type="button"
            variant="contained"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
        role="listitem"
      >
        <div
          class="c2"
        >
          <div>
            Pedreiro
          </div>
        </div>
        <div
          class=""
        >
          <button
            aria-label="Editar"
            class="MuiButtonBase-root MuiIconButton-root c3 MuiIconButton-colorPrimary"
            id="edit-test-1"
            tabindex="0"
            title="Editar"
            type="button"
            variant="contained"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Deletar"
            class="MuiButtonBase-root MuiIconButton-root c3 MuiIconButton-colorPrimary"
            id="delete-test-1"
            tabindex="0"
            title="Deletar"
            type="button"
            variant="contained"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </li>
    </ul>
  </div>
  <div
    class="Toastify"
  />
  <div
    id="modal"
  />
</DocumentFragment>
`;
